Tutorials/CNTK_209C_Image_Super-resolution_Using_CNNs_and_GANs.ipynb

```diff
@@ -0,0 +1,1562 @@
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "#gan computation graph\n",
+    "def build_GAN_graph(genG, disc, VGG, real_X_scaled, real_Y_scaled, real_Y):  \n",
+    "    #discriminator on real images\n",
+    "    D_real = discriminator(real_Y_scaled)\n",
+    "    \n",
+    "    #discriminator on fake images\n",
+    "    D_fake = D_real.clone(method = 'share', substitutions = {real_Y_scaled.output: genG.output})\n",
+    "    \n",
+    "    #VGG on real images\n",
+    "    VGG_real = VGG.clone(method = 'share', substitutions = {VGG.arguments[0]: real_Y})\n",
+    "    \n",
+    "    #VGG on fake images\n",
+    "    VGG_fake = VGG.clone(method = 'share', substitutions = {VGG.arguments[0]: 255 * genG.output})\n",
+    "    \n",
+    "    #generator loss: GAN loss + MSE loss + perceptual (VGG) loss\n",
+    "    G_loss = -square(D_fake)*0.001 + reduce_mean(square(real_Y_scaled - genG)) + reduce_mean(square(VGG_real - VGG_fake))*0.08\n",
+    "    \n",
+    "    #discriminator loss: loss on real + los on fake images\n",
+    "    D_loss = square(1.0 - D_real) + square(D_fake)\n",
+    "    \n",
+    "    G_optim = adam(G_loss.parameters,\n",
+    "                    learning_rate_schedule([(20, 0.0001), (20, 0.00001)], UnitType.minibatch, 5000),\n",
+    "                    momentum = momentum_schedule(0.9), gradient_clipping_threshold_per_sample = 0.1)\n",
+    "    \n",
+    "    D_optim = adam(D_loss.parameters,\n",
+    "                    learning_rate_schedule([(20, 0.0001), (20, 0.00001)], UnitType.minibatch, 5000),\n",
+    "                    momentum = momentum_schedule(0.9), gradient_clipping_threshold_per_sample = 0.1)\n",
+    "    \n",
+    "    G_trainer = C.Trainer(genG, (G_loss, None), G_optim)\n",
+    "    \n",
+    "    D_trainer = C.Trainer(D_real, (D_loss, None), D_optim)\n",
+    "    \n",
+    "    return (G_trainer, D_trainer)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "deletable": true,
+    "editable": true
+   },
+   "source": [
+    "### Training the model\n",
+    "<p>The train function is also considerably different now since we have to alternate updates to the discriminator and the generator. Map files and image dimensions are the same as in <b>SRResNet</b> model. Before the training, we download the pretrained <code>VGG19</code> model. It might take a few minutes.</p>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "('VGG19_ImageNet_Caffe.model', <http.client.HTTPMessage at 0x7fba3c455860>)"
+      ]
+     },
+     "execution_count": 23,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "urlretrieve(\"https://www.cntk.ai/Models/Caffe_Converted/VGG19_ImageNet_Caffe.model\", \"VGG19_ImageNet_Caffe.model\")"
```

Comment:
We are taking VGG19 model from cntk.ai location and saving it in the same path as the notebook (not deleting it after use). Notify us if this part should look different.